src/box/box.cc

```diff
@@ -793,6 +798,80 @@ wal_stream_apply_dml_row(struct wal_stream *stream, struct xrow_header *row)
 	return -1;
 }
 
+struct rlist nodes_rows[VCLOCK_MAX];
+
+/*
+ * When restoring the log is read row-by-row. However, it is necessary
+ * to store the rows for correct further recovery. For example, with
+ * mixed transactions. The following function saves the newly arrived
+ * row to the rlist, which is stored in array is nodes_rows in cell is
+ * replica_id-1. As soon as a row arrives with the is_commit flag set,
+ * the corresponding transaction tries to apply. If an error occurs the
+ * transaction is rolled back. And we continue to work with the next
+ * transaction.
+ */
+static int
+wal_stream_apply_mixed_dml_row(struct wal_stream *stream,
+			       struct xrow_header *row)
+{
+	int rc = 0;
+	/* replica_id started from 1 */
+	uint32_t id = row->replica_id - 1;
+	/* The rlist can be created if the cell of array is empty */
+	if (nodes_rows[id].prev == nodes_rows[id].next &&
+	    nodes_rows[id].prev == NULL) {
+		rlist_create(&nodes_rows[id]);
+		assert(rlist_empty(&nodes_rows[id]));
+	}
+
+	/* When a new row arrives, it has the same address. Because the
+	 * row is on the stack. Therefore, it needs to allocate memory on
+	 * the heap. We do with row->body[0].iov_base analogically.
+	 */
+	struct xrow_header *alloc_row = (struct xrow_header *)malloc(sizeof(*alloc_row));
+	if (alloc_row == NULL)
+		tnt_raise(OutOfMemory, sizeof(*alloc_row), "malloc",
+			  "alloc_row");
+	memcpy(alloc_row, row, sizeof(*row));
+
+	char *alloc_body = (char *)malloc(row->body[0].iov_len);
+	if (alloc_body == NULL)
+		tnt_raise(OutOfMemory, sizeof(*alloc_body), "malloc",
+			  "alloc_body");
+	memcpy(alloc_body, row->body[0].iov_base, row->body[0].iov_len);
+	alloc_row->body[0].iov_base = alloc_body;
```

Comment:
This seems to work just fine, so now that we've tested the hypothesis please start allocating the rows on a mempool.

Speaking of row bodies, you may use `lsregion`. For example, check applier, it had to solve exactly the same problem we have here.
Here's an example of alloc:
https://github.com/tarantool/tarantool/blob/9fcafbd14b962d9f844a9d114518fa690753f276/src/box/applier.cc#L1988-L1996

And here's an example of free:
https://github.com/tarantool/tarantool/blob/9fcafbd14b962d9f844a9d114518fa690753f276/src/box/applier.cc#L1883

The only difference is that you will need to track the oldest lsregion_id you still need to keep and free everything up to that id, not up to the id of the recently completed row.

Reply:
I'm not sure that I work correctly with `lsr_id`.